libdrgn/kallsyms.c

```diff
@@ -0,0 +1,625 @@
+	if (kind == 'u')
+		ret->binding = DRGN_SYMBOL_BINDING_UNIQUE;
+	else if (kind_lower == 'v' || kind_lower == 'w')
+		ret->binding = DRGN_SYMBOL_BINDING_WEAK;
+	else if (isupper(kind))
+		ret->binding = DRGN_SYMBOL_BINDING_GLOBAL;
+	else
+		// If lowercase, the symbol is usually local, but it's
+		// not guaranteed. Use unknown for safety here.
+		ret->binding = DRGN_SYMBOL_BINDING_UNKNOWN;
+
+	switch (kind_lower) {
+	case 'b': // bss
+	case 'c': // uninitialized data
+	case 'd': // initialized data
+	case 'g': // initialized data (small objects)
+	case 'r': // read-only data
+	case 'v': // weak object (guaranteed by elf_info() in kernel/module.c)
+		ret->kind = DRGN_SYMBOL_KIND_OBJECT;
+		break;
+	case 't': // text
+		ret->kind = DRGN_SYMBOL_KIND_FUNC;
+		break;
+	default:
+		ret->kind = DRGN_SYMBOL_KIND_UNKNOWN;
+	}
+	ret->name_lifetime = DRGN_LIFETIME_STATIC;
+}
+
+/** Compute an address via the CONFIG_KALLSYMS_ABSOLUTE_PERCPU method*/
+static uint64_t absolute_percpu(uint64_t base, int32_t val)
+{
+	if (val >= 0)
+		return (uint64_t) val;
+	else
+		return base - 1 - val;
+}
+
+/**
+ * Load the kallsyms address information from @a prog
+ *
+ * Just as symbol name loading is complex, so is address loading. Addresses may
+ * be stored directly as an array of pointers, but more commonly, they are
+ * stored as an array of 32-bit integers which are related to an offset. This
+ * function decodes the addresses into a plain array of 64-bit addresses.
+ *
+ * @param prog The program to read from
+ * @param kr The symbol registry to fill
+ * @param vi vmcoreinfo containing necessary symbols
+ * @returns NULL on success, or error
+ */
+static struct drgn_error *
+kallsyms_load_addresses(struct drgn_program *prog, struct kallsyms_reader *kr,
+			struct kallsyms_locations *loc, uint64_t **ret)
+{
+	struct drgn_error *err = NULL;
+	bool bswap, bits64;
+	_cleanup_free_ uint32_t *addr32 = NULL;
+
+	err = drgn_program_bswap(prog, &bswap);
+	if (err)
+		return err;
+	err = drgn_program_is_64_bit(prog, &bits64);
+	if (err)
+		return err;
+
+	_cleanup_free_ uint64_t *addresses =
+		malloc_array(kr->num_syms, sizeof(addresses[0]));
+	if (!addresses)
+		return &drgn_enomem;
+
+	if (loc->kallsyms_addresses) {
+		/*
+		 * The kallsyms addresses are stored as plain addresses in an
+		 * array of unsigned long! Read the appropriate size array and
+		 * do any necessary byte swaps.
+		 */
+		if (!bits64) {
+			addr32 = malloc_array(kr->num_syms, sizeof(addr32[0]));
+			if (!addr32)
+				return &drgn_enomem;
+
+			err = drgn_program_read_memory(prog, addr32,
+						loc->kallsyms_addresses,
+						kr->num_syms * sizeof(uint32_t),
+						false);
+			if (err)
+				return err;
+			for (int i = 0; i < kr->num_syms; i++) {
+				if (bswap)
+					addresses[i] = bswap_32(addr32[i]);
+				else
+					addresses[i] = addr32[i];
+			}
+		} else {
+			err = drgn_program_read_memory(prog, addresses,
+						loc->kallsyms_addresses,
+						kr->num_syms * sizeof(uint32_t),
```

Comment:
Is this supposed to be `sizeof(uint64_t)`?

Reply:
Yes, what a catch. I'm not sure how I didn't notice the broken addresses for these.

I'm able to test:

- `!BASE_RELATIVE && !ABSOLUTE_PERCPU` - this means kallsyms addresses are just an array of unsigned long. This is _not_ the default, but it's possible to build a normal x86_64 kernel with this. This was affected by this issue and will be fixed.
- `BASE_RELATIVE && ABSOLUTE_PERCPU` - this means kallsyms addresses are 32-bit with a base address, and they have special handling for negative values for percpu region. This is the default for all x86_64 kernels.
- `BASE_RELATIVE && !ABSOLUTE_PERCPU` - same as the previous, but no handling of negatives for the percpu region. Most x86_64 kernels fail to build because they can't encode percpu addresses. But this is the default for aarch64 kernels and so it gets lots of testing too.

There is no `!BASE_RELATIVE && ABSOLUTE_PERCPU` - or maybe there is but it doesn't result in a functional difference, so there's nothing to test.